qemu/tests/migration_with_json_backing.py

```diff
@@ -0,0 +1,131 @@
+            msg = "Create image by command: %s" % qemu_img_cmd
+            LOG_JOB.info(msg)
+            cmd_result = process.run(
+                qemu_img_cmd, shell=True, verbose=False, ignore_status=True)
+
+    def src_sn_chain_setup(images_tag):
+        images = get_img_objs(images_tag)
+        for image in images[1:len(images)-1]:
+            _create_image_with_backing(image)
+        return images
+
+    def blockdev_add_image(tag, is_backing_null=False):
+        image_params = params.object_params(tag)
+        devices = vm.devices.images_define_by_params(tag,
+                                                     image_params,
+                                                     'disk')
+        devices.pop()
+        for dev in devices:
+            if vm.devices.get_by_qid(dev.get_qid()):
+                continue
+            if isinstance(dev, QBlockdevFormatNode) and is_backing_null:
+                dev.params["backing"] = None
+            ret = vm.devices.simple_hotplug(dev, vm.monitor)
+            if not ret[1]:
+                test.fail("Failed to hotplug '%s': %s."
+                          % (dev, ret[0]))
+
+    def create_snapshot():
+        options = ["node", "overlay"]
+        cmd = "blockdev-snapshot"
+        arguments = params.copy_from_keys(options)
+        arguments.setdefault("node", "drive_%s" % params["base_tag"])
+        arguments.setdefault("overlay", "drive_%s" % params["snapshot_tag"])
+        return vm.monitor.cmd(cmd, dict(arguments))
+
+    def verify_backing_chain(info):
+        """Verify image's backing chain."""
+        for image, img_info in zip(images, reversed(info)):
+            base_image = None
+            if image.base_tag:
+                base_params = params.object_params(image.base_tag)
+                base_image = get_image_repr(image.base_tag,
+                                            base_params,
+                                            get_data_dir(),
+                                            'filename')
+            backing_info = img_info.get("full-backing-filename")
+            if backing_info and "json" in backing_info:
+                back_info = backing_info.strip("json:")
+                backing_info = json.loads(back_info)["file"]["filename"]
+            if base_image != backing_info:
+                test.fail(("backing chain check for image %s failed, backing"
+                           " file from info is %s, which should be %s.") %
+                          (image.image_filename, backing_info,
+                           base_image))
+
+    def check_backing_file(image):
+        out = json.loads(image.info(force_share=True, output="json"))
+        verify_backing_chain(out)
+
+    def clean_images(images):
+        for image in images:
+            image.remove()
+
+    images_tag = params.get("image_chain").split()
+    params["image_name_%s" % images_tag[0]] = params["image_name"]
+    try:
+        images = src_sn_chain_setup(images_tag)
```

Comment:
if we want to throw an exception when hitting an error, images here may not be defined, in finally clause, we may get an undefined error, so what about move line55 before try, then images contains the snapshot image objects, then src_sn_chain_setup(images), no return.

In this case, even an exception occurs, clean_images(images[1:]) can work fine because image.remove will only remove an existing file.